File: src/chipschallenge/blockreactions/BlueWallFakeTo.java
package chipschallenge.blockreactions;

import chipschallenge.Block;
import chipschallenge.Block.Type;
import chipschallenge.BlockContainerFullException;

/**
 * Move to Blue fake wall
 */
public class BlueWallFakeTo extends BlockReaction {

    private BlueWallFakeTo() {
    }
    private static BlueWallFakeTo mInstance = null;

    public static synchronized BlueWallFakeTo getInstance() {
        if (mInstance == null) {
            mInstance = new BlueWallFakeTo();
        }
        return mInstance;
    }

    // Wall is converted to floor
    public void react(Block moving, Block standing) throws BlockContainerFullException {
        standing.replace(createBlock(Type.FLOOR));
    }

    // Only chip can move to it
    public boolean canMove(Block moving, Block standing) {
        return moving.isChip();
    }
}

File: ChipsChallenge.Shared/Blockreactions/BlueWallFakeTo.cs
namespace ChipsChallenge.Shared.Blockreactions
{
    using Type = Block.Type;

    public class BlueWallFakeTo : NoSlipReaction
    {
        private static BlueWallFakeTo instance;

        private BlueWallFakeTo()
        {
        }

        public static BlueWallFakeTo Instance
        {
            get
            {
                lock (typeof(BlueWallFakeTo))
                {
                    return instance ?? (instance = new BlueWallFakeTo());
                }
            }
        }

        public override void React(Block moving, Block standing)
        {
            // Wall is converted to floor
            standing.Replace(CreateBlock(Type.FLOOR));
        }

        public override bool canMove(Block moving, Block standing)
        {
            // Only chip can move to it
            return moving.Chip;
        }
    }
}